### ISSUE TITLE
"Bind" cannot be used with transparent support enabled.

### ISSUE THREAD

**Issue by vagabondking** [tinyproxy/tinyproxy#15]
I've got 2 interfaces and I'm trying to ensure the outbound traffic goes through a specific interface.   I found a few references to this online but most of them pointed to your old mailing list or bugzilla which are offline, so I could not research this.   Is there a specific reason that Bind cannot be used with transparent support?

tinyproxy 1.8.4

**Comment by falk0069** [comment]
Had the same problem as you.  I came across an openwrt fix here: <br>
`https://gitlab.labs.nic.cz/turris/openwrt-packages/commit/7c01da4a72e6f0b7613a86529547659ea4007eba`

So, instead of installing via yum/apt I did a git clone.  Here were the steps I did which may vary for you:

```
sudo su -
git clone https://github.com/tinyproxy/tinyproxy.git
cd tinyproxy
touch ChangeLog  #odd error if this file doesn't exist)
autoreconf --install --symlink -f
yum install asciidoc  #dependency I was missing
./configure
make
```
if it built successfully you are home free.  Next edit this file:
`./src/conf.c`

delete the lines prefixed with --
```
--#ifndef TRANSPARENT_PROXY
        int r = set_string_arg (&conf->bind_address, line, &match[2]);

        if (r)
                return r;
        log_message (LOG_INFO,
                     "Outgoing connections bound to IP %s", conf->bind_address);
        return 0;
--#else
--        fprintf (stderr,
--                 "\"Bind\" cannot be used with transparent support enabled.\n");
--        return 1;
--#endif
}
```
then build and deploy it:
```
make
make install
```

The default installation location is /usr/local (which you can change with the configure command)  So, the config file by default will be here:
`/usr/local/etc/tinyproxy/tinyproxy.conf`

And the binary here:
`/usr/local/sbin/tinyproxy`

You may also need to create the log and run location
```
mkdir -p /usr/local/var/log/tinyproxy/
mkdir -p /usr/local/var/run/tinyproxy/
chmod 777 /usr/local/var/log/tinyproxy/
chmod 777 /usr/local/var/run/tinyproxy/
```

This got me up and going and allowed for the Bind option to actually work.  I'm going to be changing the install directory once I figure out if this is the solution for me.  Hopefully there is a more official one, but I thought I'd share this in case others find it useful.

**Comment by jeremyherbert** [comment]
Is there a reason this bug isn't fixed? This has been a problem in the debian package since 2011...

**Comment by sebastianst** [comment]
@jeremyherbert Please do make a PR.

**Comment by ppanero** [comment]
@sebastianst https://github.com/tinyproxy/tinyproxy/pull/97

**Comment by obnoxxx** [comment]
There must have been a reason to disable bind when transparent mode is enabled.... Not sure any more. :-)

**Comment by obnoxxx** [comment]
Not sure. Maybe it's just that binding to a specific server address somehow contradicts the stealth mode of a transparent proxy?

**Comment by obnoxxx** [comment]
fixed via #97

**Comment by rofl0r** [comment]
actually not. the patch from @ppanero is broken and just removes the error message, but doesn't actually bind the interface. further it invokes UB by not returning anything at all (the function is supposed to return an int).
i found this due to a warning issued by GCC due about missing return value, after i enabled transparent proxy support by default in #147 .
a fix is in the works and will be committed as a new PR in a couple minutes.
